opm/simulators/wells/GroupState.cpp

```diff
@@ -419,6 +423,63 @@ has_gpmaint_target(const std::string& gname) const
     return (this->m_gpmaint_target.count(gname) > 0);
 }
 
+template<class Scalar>
+void GroupState<Scalar>::
+update_gconsump(const Schedule& schedule, const int report_step, const SummaryState& summary_state) {
+    this->m_gconsump_rates.clear();
+    const auto& sched_state = schedule[report_step];
+    const auto& gconsump = sched_state.gconsump();
+    auto gcr_recursive =
+        [this, &sched_state, &gconsump, &summary_state](auto self,
+                                                        const std::string& group_name,
+                                                        std::pair<Scalar, Scalar>& rates,
+                                                        const double parent_gefac = 1.0) -> bool
+        {
+            // If group already has been computed, update parent rates and return true
+            const auto it = this->m_gconsump_rates.find(group_name);
+            if (it != this->m_gconsump_rates.end()) {
+                rates.first += static_cast<Scalar>(it->second.first * parent_gefac);
+                rates.second += static_cast<Scalar>(it->second.second * parent_gefac);
+                return true;
+            }
+
+            // Accumulate from sub-groups and keep track of any updates in 'has_values'
+            bool has_values = false;
+            if (sched_state.groups.has(group_name)) {
+                for (const auto& child_gname : sched_state.groups(group_name).groups()) {
+                    const auto gefac = sched_state.groups(child_gname).getGroupEfficiencyFactor();
+                    has_values = has_values || self(self, child_gname, rates, gefac);
+                }
+            }
+
+            // Add consumption/import rates at current level
+            if (gconsump.has(group_name)) {
+                const auto& group_gc = gconsump.get(group_name, summary_state);
+                rates.first += static_cast<Scalar>(group_gc.consumption_rate);
+                rates.second += static_cast<Scalar>(group_gc.import_rate);
+                has_values = true;
+            }
+
+            // Update map if values are set
+            if (has_values) this->m_gconsump_rates[group_name] = rates;
+            return has_values;
+        };
+
+    std::pair<Scalar, Scalar> rates = {0.0, 0.0};
```

Comment:
This will generate a type conversion if `Scalar` is different from `double`.  I think I'd personally use class template argument deduction and write it as
```C++
auto rates = std::pair { Scalar{0}, Scalar{0} };
```